src/SubHask/Algebra/Vector.hs

```diff
@@ -1538,104 +1531,181 @@ type instance SetElem (ACCVector (bknd::Backend) n r) b = ACCVector (bknd::Backe
     {-# INLINE (.*)   #-}
     (.*) (ACCVector  v) r = ACCVector (A.map (\x -> x P.* (A.constant r)) v)
 
-instance (VectorSpace r, ValidACCVector bknd n r, IsScalar r, A.IsFloating r) => VectorSpace (ACCVector (bknd::Backend) (n::Symbol) r) where
+instance (P.Fractional (A.Exp r), VectorSpace r, ValidACCVector bknd n r, IsScalar r, A.IsFloating r) => VectorSpace (ACCVector (bknd::Backend) (n::Symbol) r) where
     {-# INLINE (./)   #-}
     (./) (ACCVector  v) r = ACCVector (A.map (\x -> x P./(A.constant r)) v)
 
     {-# INLINE (./.)  #-}
     (./.) (ACCVector a1) (ACCVector a2) = ACCVector (A.zipWith (P./) a1 a2)
 
 
-newtype ACCMatrix (bknd::Backend) (m::k1) (n::k2) a = ACCMatrix (A.Acc (A.Array A.DIM2 a))
+instance (P.Num (A.Exp r), FreeModule r, ValidLogic r, ValidACCVector b n r, IsScalar r) => FiniteModule (ACCVector b (n::Symbol) r)
+
+-- this returns an A.Exp Int and not an Int, which makes GHC complain
+-- where
+--
+--     {-# INLINE dim #-}
+--     dim (ACCVector v) = A.length v
 
-data ACCMatrix' b (m::k1) (n::k2) r where
-  Id    :: {-#UNPACK#-}!r              -> ACCMatrix' b m m r
-  Diag  :: {-#UNPACK#-}!(ACCVector b m r)  -> ACCMatrix' b m m r
-  Mat   :: {-#UNPACK#-}!(ACCMatrix b m n r)  -> ACCMatrix' b m n r
 
 
-type instance Scalar (ACCMatrix' b r m n) = Scalar r
-type instance (ACCMatrix' b r m n)><r = ACCMatrix b r m n
+instance
+    ( P.Num (A.Exp r)
+    , Monoid r
+    , ValidLogic r
+    , ValidACCVector b n r
+    , A.Elt r
+    , IsScalar r
+    , FreeModule r
+    ) => IxContainer (ACCVector b (n::Symbol) r)
+        where
+    --(!) also returns A.Exp r where r is expected ...
+    -- {-# INLINE (!) #-}
+    -- (!) (ACCVector v) i =  v A.! (A.index1 (A.lift i))
+    -- {-# INLINABLE  toIxList #-}
+    -- toIxList v = P.zip [0..] $ go (dim v-1) []
+    --     where
+    --         go (-1) xs = xs
+    --         go    i xs = go (i-1) (v!i : xs)
+    -- fails on f with:
+    -- Couldn't match type ‘Int’ with ‘A.Exp A.DIM1’
+    -- Expected type: A.Exp A.DIM1 -> A.Exp r -> A.Exp b1
+    --   Actual type: Index (ACCVector b n r)
+    --                -> Elem (ACCVector b n r) -> b1
+    -- {-# INLINABLE imap #-}
+    -- imap f (ACCVector v) = ACCVector (A.zipWith f (A.generate (A.shape v) P.id) v)
+
+    type ValidElem (ACCVector b n r) e = (ClassicalLogic e, IsScalar e, FiniteModule e, ValidACCVector b n e)
+
+
+instance (Eq r, Monoid r, ValidACCVector b n r) => Eq_ (ACCVector b (n::Symbol) r) where
+    -- returns A.Exp Bool . . . .
+    -- {-# INLINE (==) #-}
+    -- (ACCVector v2) == (ACCVector v1) = (A.lift v1) A.==* (A.lift v2)
+
+--Will acc arrays support Logic r~Bool?
+instance
+    ( ValidACCVector b n r
+    , P.Num (A.Exp r)
+    , ExpField r
+    , Normed r
+    , Ord_ r
+    , Logic r~Bool
+    , IsScalar r
+    , VectorSpace r
+    ) => Metric (ACCVector b (n::Symbol) r)
+        -- where
+
+    -- also seems to wan to deduce r ~ A.Acc (Scalar r)
```

Comment:
I'm guessing this has more to do with accelerate than with subhask, but I'm not sure.